test/run_test.py

```diff
@@ -460,6 +599,14 @@ def main():
     if options.jit:
         selected_tests = filter(lambda test_name: "jit" in test_name, TESTS)
 
+    if options.determine_from is not None and os.path.exists(options.determine_from):
+        # HACK: Ensure the 'test' paths can be traversed by Modulefinder
+        sys.path.append('test')
```

Comment:
I am... pretty sure this isn't going to work for nested tests, e.g., `test/distributed`

Reply:
> Overall: This is neat, but it is a kind of chunky piece of code that, if done wrong, could lead to us missing CI signal on PRs. So it would be good to get a clearer characterization of what benefits we'll get when this gets merged.

I just ran an analysis over the most recent 100 master commits (I'm in the process of turning that into a unit test to assuage fears of under-testing). Out of the 21 "slow tests", this determination would have run a mean of 17 tests per commit. test_nn would run on ~80% of the commits. Determination will only run on PRs and not on master, but the makeup of master commits is probably similar to that of PR commits.